### PR TITLE
Plugin subsystem. Shown as a last option in song context menu. 

### PR DESCRIPTION
**Pull request by Adonai**
Part of #407

### CODE REVIEW

File: AndroidManifest.xml [hunk 1]
```diff
@@ -127,13 +127,15 @@ THE SOFTWARE.
...
+		<service android:name="PlaybackService" android:exported="true">
 			<intent-filter>
 				<action android:name="ch.blinkenlights.android.vanilla.action.PLAY" />
 				<action android:name="ch.blinkenlights.android.vanilla.action.PAUSE" />
 				<action android:name="ch.blinkenlights.android.vanilla.action.TOGGLE_PLAYBACK" />
 				<action android:name="ch.blinkenlights.android.vanilla.action.NEXT_SONG" />
 				<action android:name="ch.blinkenlights.android.vanilla.action.PREVIOUS_SONG" />
+				<!-- accepting plugin answers on their desc query -->
+				<action android:name="com.vanilla.action.HANDLE_PLUGIN_PARAMS" />
 			</intent-filter>
```
Comment (Adonai) [inline AndroidManifest.xml]: Original one didn't fit into 120 lines :(
Ok, I fixed it

File: src/ch/blinkenlights/android/vanilla/FileUtils.java [hunk 2]
```diff
@@ -47,15 +47,15 @@ public static boolean canDispatchIntent(Intent intent) {
...
 			try {
 				URI uri = new URI("file", path, null);
 				String mimeGuess = URLConnection.guessContentTypeFromName(uri.toString());
 				if (mimeGuess != null && mimeGuess.matches("^(image|text)/.+")) {
 					canDispatch = true;
 				}
 			} catch (URISyntaxException e) {
-				Log.e("VanillaMusic", "failed to encode "+path+": "+e);
+				Log.e("VanillaMusic", "failed to encode " + path + ": " + e);
 			}
```
Comment (Adonai) [inline src/ch/blinkenlights/android/vanilla/FileUtils.java]: done, removed

File: src/ch/blinkenlights/android/vanilla/LibraryActivity.java [hunk 3]
```diff
@@ -139,6 +143,10 @@
 	 */
 	private long mLastActedId;
 	/**
+	 * Holds last intent that was passed to the context menu
+	 */
+	private Intent lastRequestedCtx;
+	/**
```
Comment (adrian-bl) [inline src/ch/blinkenlights/android/vanilla/LibraryActivity.java]: This should be mLastRequestedCtx, see http://source.android.com/source/code-style.html#follow-field-naming-conventions

File: src/ch/blinkenlights/android/vanilla/PlaybackService.java [hunk 4]
```diff
@@ -1579,8 +1592,24 @@ public boolean handleMessage(Message message)
...
+		case MSG_HANDLE_PLUGIN:
+			// reset plugin collection timer
+			mHandler.removeMessages(MSG_DELAY_PLUGIN_COLLECTION);
 
+			Intent descIntent = (Intent) message.obj;
+			String pluginName = descIntent.getStringExtra(PluginUtils.EXTRA_PARAM_PLUGIN_NAME);
+			ApplicationInfo info = descIntent.getParcelableExtra(PluginUtils.EXTRA_PARAM_PLUGIN_APP);
+			mPlugins.put(pluginName, info);
 
+			mHandler.sendEmptyMessageDelayed(MSG_DELAY_PLUGIN_COLLECTION, 200);
```
Comment (adrian-bl) [inline src/ch/blinkenlights/android/vanilla/PlaybackService.java]: Where does the 200ms come from?
Wouldn't it be possible to do this without a magic timeout? (eg: query the package manager how many plugins are actually installed and check if all of them replied?)

Comment (Adonai) [inline src/ch/blinkenlights/android/vanilla/PlaybackService.java]: Thought about it. But what if plugin won't answer in time? We'd wait forever for it.

Comment (Adonai) [inline src/ch/blinkenlights/android/vanilla/PlaybackService.java]: Ok, how about that: we give all plugins 1 second timeout. If everybody replied, then we fire the dialog right away. If not, we wait this 1 second and if nothing is received, we publish list with the ones that were kind enough to reply.

File: src/ch/blinkenlights/android/vanilla/LibraryActivity.java (hunk 3 above)
Comment (Adonai) [inline src/ch/blinkenlights/android/vanilla/LibraryActivity.java]: Done

File: src/ch/blinkenlights/android/vanilla/PlaybackService.java (hunk 4 above)
Comment (adrian-bl) [inline src/ch/blinkenlights/android/vanilla/PlaybackService.java]: Yes, that would sound like a good plan

Comment (Adonai) [inline src/ch/blinkenlights/android/vanilla/PlaybackService.java]: Done

**Comment by Adonai** [comment]
Shall I continue to work on this? As I see this work was in vain as we'll integrate ID3/VC scanner into Vanilla, according to #460, so there's no need for plugins for tag editing or lyrics viewing from metadata anymore, this can all just be implemented as a dialog.

**Comment by adrian-bl** [comment]
> As I see this work was in vain as we'll integrate ID3/VC scanner into Vanilla, according to #460, so 
> there's no need for plugins for tag editing or lyrics viewing from metadata anymore, this can all just be 
> implemented as a dialog.

This shouldn't really affect the *editor* - the new media scanner is supposed to pick up changes just like the system-one does - so the plugin should not be affected by the new media library

**Comment by Adonai** [comment]
@adrian-bl, Okay, will have some time to work on it this week

**Comment by Adonai** [comment]
@adrian-bl , still don't get it. The old proposal was to incorporate ContentResolver calls here. Now that I see all these calls being ripped off in main repo, what shall I do in the plugins?

**Comment by adrian-bl** [comment]
The ContentResolver calls itself did nothing to pick up changes: This was done by the ContentObserver which will return to the MediaLibrary when i have time to implement it :-)

The plan is to have the mediascanner listening to changes (like the playback service did in the old days) and scan for modified files after a change was broadcasted (which is easy as the media table has an mtime row).

So in the end nothing should change for the plugin system: Modified files are 'discovered' automatically

**Comment by Adonai** [comment]
So I'm to implement ContentResolver calls with custom URL in plugins, which will resolve to our new media library row, is that correct?

**Comment by Adonai** [comment]
Done, all work now resides in LibraryActivity. It now registers/unregisters broadcast receiver for plugin intent handling.

File: src/ch/blinkenlights/android/vanilla/LibraryActivity.java [hunk 5]
```diff
@@ -140,6 +149,19 @@
...
+	private Intent mLastRequestedCtx;
+	/**
+	 * Plugin descriptions and packages.
+	 * Keys are plugin names, values are their packages
+	 */
+	private Map<String, ApplicationInfo> mPlugins = new HashMap<>();
+	/**
+	 * Broadcast receiver for plugin collecting
+	 */
+	private BroadcastReceiver pluginInfoReceiver;
```
Comment (adrian-bl) [inline src/ch/blinkenlights/android/vanilla/LibraryActivity.java]: that should be mPluginInfoReceiver

File: src/ch/blinkenlights/android/vanilla/LibraryActivity.java [hunk 6]
```diff
@@ -195,6 +218,37 @@ public void onCreate(Bundle state)
 		bindControlButtons();
 	}
 
+	private void showPluginMenu() {
```
Comment (adrian-bl) [inline src/ch/blinkenlights/android/vanilla/LibraryActivity.java]: Please add the standard boilerplate comment on top of the function (@desc, etc)

**Comment by adrian-bl** [comment]
Thanks: i added two comments (just some style / comment stuff) - i'll merge the code after this has been fixed :-)

**Comment by Adonai** [comment]
@adrian-bl , done, thanks! After months of absence I totally forgot Android code style ;)

**Comment by adrian-bl** [comment]
Thanks! I merged it and triggered a nightly build.

Btw: I noticed that the tag-editor behaves strange: Somehow the tags of the first edited item seem to 'stick' to it: i always get the same tags presented, no matter which songs i select to edit.

(The Lyrics plugin seems to work fine)

**Comment by Adonai** [comment]
@adrian-bl , will take a look and mend to it, thanks.
I'll initiate the process of submitting plugins to F-Droid and GPlay

**Comment by Adonai** [comment]
@adrian-bl , fixed that

**Comment by adrian-bl** [comment]
About the fdroid/play releases: keep in mind that it will take some time until i publish the next release as there are still a few issues with the media db.

I expect to push an fdroid release at the end of jan. Google Play will probably follow a month later.

**Comment by Adonai** [comment]
I know, I'll just prepare needed merge-requests, drafts and screenshots for both systems.

**Pull request merged**